### PR TITLE
Dashboard Maps PoC

### PR DESCRIPTION
**Pull request by czue**
[Context / Working Spec Here](https://docs.google.com/document/d/19Ds59vZ7PHRcNIEFjOoIQBIJSZ3r8t9idR-wZdr-SkM/edit)

This PR has a few pieces that I used to make a little map PoC in the dashboard:

1. A test data generator that generates random visits.
2. A SQL query (based on the one used in superset) to format the map data.
3. A view to show the data on a map.

This is the end result:

[connect-map-poc.webm](https://github.com/user-attachments/assets/f50cf336-34fa-4895-b145-1fab460c56c8)

Of course the data looks ridiculous when randomly generated, which is one reason it would be nice to have this deployed to the real app.

### CODE REVIEW

**Review by calellowitz — COMMENTED** [review]
A couple small  things but generally looks good. Nothing blocking, but curious to hear a bit more. Also curious what the motivation for raw SQL over the ORM was. Was it about trying push that logic to postgres more easily, or just about what you were starting with?

File: commcare_connect/reports/urls.py [hunk 1]
```diff
@@ -6,4 +6,6 @@
 
 urlpatterns = [
     path("delivery_stats", views.delivery_stats_report, name="delivery_stats_report"),
+    path("program_dashboard", views.program_dashboard_report, name="program_dashboard_report"),
+    path("api/visit-map-data/", views.visit_map_data, name="visit_map_data"),
```
[the remark this answers is not in the document]
Comment (czue) [reply commcare_connect/reports/urls.py]: oops, thanks. https://github.com/dimagi/commcare-connect/pull/403/commits/a23b1688e29e8f59cf77ab79b83fd1c3ff50646b

File: commcare_connect/reports/sql/visit_map.sql [hunk 2]
```diff
@@ -0,0 +1,38 @@
+SELECT
+	ou.opportunity_id,
+	ou.xform_id AS visit_id,
+	ou.visit_date,
+	ou.form_json -> 'metadata' -> 'username' AS username_connectid,
+	COALESCE(ou.form_json -> 'form' -> 'case' -> 'update' ->> 'visit_type', 'unknown') AS visit_type,
```
[the remark this answers is not in the document]
Comment (czue) [reply commcare_connect/reports/sql/visit_map.sql]: good call, removed in https://github.com/dimagi/commcare-connect/pull/403/commits/10b2a3833dc7578317d35148875758bc75b93f4a

File: commcare_connect/reports/sql/visit_map.sql [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	ou.reason AS rejection_reason,
+	ou.form_json -> 'metadata' -> 'timeStart' AS timeStart,
+	ou.form_json -> 'metadata' -> 'timeEnd' AS timeEnd,
+	EXTRACT(EPOCH FROM ((ou.form_json -> 'metadata' ->> 'timeEnd')::timestamp - (ou.form_json -> 'metadata' ->> 'timeStart')::timestamp)) / 60 AS visit_duration,
+
+	CASE
+	WHEN NULLIF(ou.form_json -> 'form' -> 'case' -> 'update' ->> 'confirm_practice_case', '') IS NOT NULL
+	THEN 'yes'
+	ELSE NULL
+  END AS practice_case,
```
[the remark this answers is not in the document]
Comment (czue) [reply commcare_connect/reports/sql/visit_map.sql]: removed in https://github.com/dimagi/commcare-connect/pull/403/commits/10b2a3833dc7578317d35148875758bc75b93f4a

File: commcare_connect/templates/reports/dashboard.html [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        style: 'mapbox://styles/mapbox/dark-v11',
+        center: [-103.5917, 40.6699],
+        zoom: 3
+    });
+
+    map.on('load', () => {
+        // Add a new source from our GeoJSON data and
+        // set the 'cluster' option to true. GL-JS will
+        // add the point_count property to your source data.
+        map.addSource('earthquakes', {
```
[the remark this answers is not in the document]
Comment (czue) [reply commcare_connect/templates/reports/dashboard.html]: https://github.com/dimagi/commcare-connect/pull/403/commits/3691df338e0593eed8fe89535303cedc47b68411

**Comment by czue** [comment]
> Also curious what the motivation for raw SQL over the ORM was. Was it about trying push that logic to postgres more easily, or just about what you were starting with?

Bit of both probably more the latter. When I saw how superset was extracting the lat/lon from the form JSON I figured it'd be easier to just copy it across. I don't (yet) have an opinion about whether ORM or raw SQL will be better / more maintainable long term. So given that, I guess I'd say do the easier thing. But happy to port it to the ORM now if you think that's obviously the right long term choice.

**Comment by czue** [comment]
> Do we need all these columns? It looks like most aren't used.

We definitely don't use them yet, but I figured if someone put them in superset we'll probably want them in the dashboard soon enough. But I can delete and re-add the ones that actually come up if you prefer

**Comment by czue** [comment]
Re: SQL vs ORM, [if materialized views are a pattern this project is going to adopt](https://github.com/dimagi/commcare-connect/pull/404/) that might end up being the best of all worlds.

**Review by calellowitz — APPROVED** [review]
Looks good, thanks. I think the view point is an interesting one, but this is early enough, that I am fine with the raw sql unless you can see other uses of this query, in which case being able to treat it like a django model would be nice. The other nice thing about views is they can be accessed in superset directly, since they are part of the schema, which would probably make that team's life a bit easier

**Comment by czue** [comment]
> The other nice thing about views is they can be accessed in superset directly, since they are part of the schema, which would probably make that team's life a bit easier

Ah nice, definitely makes sense to put in a view then at some point I'd say once the dust has settled some.

**Pull request merged**